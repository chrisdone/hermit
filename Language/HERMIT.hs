module Language.HERMIT where

version = "0.0"
