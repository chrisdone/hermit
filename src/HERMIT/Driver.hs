module HERMIT.Driver where

ghcFlags :: [String]
ghcFlags = [ "-fforce-recomp"
           , "-O2"
           , "-dcore-lint"
           , "-fsimple-list-literals"
           , "-fexpose-all-unfoldings"
--           , "-v0"
           ]
