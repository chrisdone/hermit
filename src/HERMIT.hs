module HERMIT (plugin) where

import Data.List (isPrefixOf)
import Data.Maybe (fromMaybe)

import HERMIT.GHC
import HERMIT.Optimize
import HERMIT.Plugin (getPhaseFlag)

plugin :: Plugin
plugin = optimize $ \ options -> let (pn,opts) = fromMaybe (0,options) (getPhaseFlag options)
                                 in phase pn $ interactive [] opts
