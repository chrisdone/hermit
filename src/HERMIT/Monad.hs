{-# LANGUAGE CPP, InstanceSigs #-}

module HERMIT.Monad
          (
            -- * The HERMIT Monad
            HermitM
          , runHM
          , liftCoreM
          , newGlobalIdH
          , newIdH
          , newTyVarH
          , newCoVarH
          , cloneVarH
            -- * Saving Definitions
          , Label
          , DefStash
          , saveDef
          , lookupDef
          , getStash
            -- * Messages
          , HermitMEnv(..)
          , DebugMessage(..)
          , mkHermitMEnv
          , sendDebugMessage
) where

import Prelude hiding (lookup)

import Data.Map

import Control.Monad
import Control.Monad.IO.Class
import Control.Applicative
import Control.Arrow

import Language.KURE

import HERMIT.Core
import HERMIT.Context
import HERMIT.Kure.SumTypes
import HERMIT.GHC

----------------------------------------------------------------------------

-- | A label for individual definitions.
type Label = String

-- | A store of saved definitions.
type DefStash = Map Label CoreDef

-- | A way of sending messages to top level
newtype HermitMEnv = HermitMEnv { hs_debugChan :: DebugMessage -> HermitM () }

-- | The HERMIT monad is kept abstract.
newtype HermitM a = HermitM (HermitMEnv -> DefStash -> CoreM (KureM (DefStash, a)))

runHermitM :: HermitM a -> HermitMEnv -> DefStash -> CoreM (KureM (DefStash, a))
runHermitM (HermitM f) = f

-- | Eliminator for 'HermitM'.
runHM :: HermitMEnv -> DefStash -> (DefStash -> a -> CoreM b) -> (String -> CoreM b) -> HermitM a -> CoreM b
runHM env s success failure ma = runHermitM ma env s >>= runKureM (\ (a,b) -> success a b) failure

----------------------------------------------------------------------------

-- | Get the stash of saved definitions.
getStash :: HermitM DefStash
getStash = HermitM (\ _ s -> return $ return (s, s))

-- | Replace the stash of saved definitions.
putStash :: DefStash -> HermitM ()
putStash s = HermitM (\ _ _ -> return $ return (s, ()))

sendDebugMessage :: DebugMessage -> HermitM ()
sendDebugMessage msg = do env <- HermitM $ \ ch s -> return $ return (s, ch)
                          hs_debugChan env msg

-- | Save a definition for future use.
saveDef :: Label -> CoreDef -> HermitM ()
saveDef l d = getStash >>= (insert l d >>> putStash)

-- | Lookup a previously saved definition.
lookupDef :: Label -> HermitM CoreDef
lookupDef l = getStash >>= (lookup l >>> maybe (fail "Definition not found.") return)

----------------------------------------------------------------------------

instance Functor HermitM where
  fmap :: (a -> b) -> HermitM a -> HermitM b
  fmap = liftM

instance Applicative HermitM where
  pure :: a -> HermitM a
  pure = return

  (<*>) :: HermitM (a -> b) -> HermitM a -> HermitM b
  (<*>) = ap

instance Monad HermitM where
  return :: a -> HermitM a
  return a = HermitM $ \ _ s -> return (return (s,a))

  (>>=) :: HermitM a -> (a -> HermitM b) -> HermitM b
  (HermitM gcm) >>= f = HermitM $ \ env -> gcm env >=> runKureM (\ (s', a) -> runHermitM (f a) env s') (return . fail)

  fail :: String -> HermitM a
  fail msg = HermitM $ \ _ _ -> return (fail msg)

instance MonadCatch HermitM where
  catchM :: HermitM a -> (String -> HermitM a) -> HermitM a
  (HermitM gcm) `catchM` f = HermitM $ \ env s -> gcm env s >>= runKureM (return.return) (\ msg -> runHermitM (f msg) env s)

----------------------------------------------------------------------------

-- | 'CoreM' can be lifted to 'HermitM'.
liftCoreM :: CoreM a -> HermitM a
liftCoreM ma = HermitM $ \ _ s -> do a <- ma
                                     return (return (s,a))

instance MonadIO HermitM where
  liftIO :: IO a -> HermitM a
  liftIO = liftCoreM . liftIO

instance MonadUnique HermitM where
  getUniqueSupplyM :: HermitM UniqSupply
  getUniqueSupplyM = liftCoreM getUniqueSupplyM

instance MonadThings HermitM where
  lookupThing :: Name -> HermitM TyThing
  lookupThing = liftCoreM . lookupThing

instance HasDynFlags HermitM where
  getDynFlags :: HermitM DynFlags
  getDynFlags = liftCoreM getDynFlags

----------------------------------------------------------------------------

newName :: String -> HermitM Name
newName nm = mkSystemVarName <$> getUniqueM <*> pure (mkFastString nm)

-- | Make a unique global identifier for a specified type, using a provided name.
newGlobalIdH :: String -> Type -> HermitM Id
newGlobalIdH nm ty = mkVanillaGlobal <$> newName nm <*> pure ty

-- | Make a unique identifier for a specified type, using a provided name.
newIdH :: String -> Type -> HermitM Id
newIdH nm ty = mkLocalId <$> newName nm <*> pure ty

-- | Make a unique type variable for a specified kind, using a provided name.
newTyVarH :: String -> Kind -> HermitM TyVar
newTyVarH nm k = mkTyVar <$> newName nm <*> pure k

-- | Make a unique coercion variable for a specified type, using a provided name.
newCoVarH :: String -> Type -> HermitM TyVar
newCoVarH nm ty = mkCoVar <$> newName nm <*> pure ty

-- | Make a new variable of the same type, with a modified textual name.
cloneVarH :: (String -> String) -> Var -> HermitM Var
cloneVarH nameMod v | isTyVar v = newTyVarH name ty
                    | isCoVar v = newCoVarH name ty
                    | isId v    = newIdH name ty
                    | otherwise = fail "If this variable isn't a type, coercion or identifier, then what is it?"
  where
    name = nameMod (uqName v)
    ty   = varType v

----------------------------------------------------------------------------

-- | A message packet.
data DebugMessage = DebugTick String
                  | DebugCore String HermitC CoreTC

mkHermitMEnv :: (DebugMessage -> HermitM ()) -> HermitMEnv
mkHermitMEnv debugger = HermitMEnv { hs_debugChan = debugger }

----------------------------------------------------------------------------
