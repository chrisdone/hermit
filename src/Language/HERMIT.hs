{-# LANGUAGE KindSignatures, GADTs, TemplateHaskell #-}
module Language.HERMIT
        ( DSL.H                     -- abstact
        , DSL.apply                 -- DSL
        , DSL.consider              -- DSL
        ) where

import Language.HERMIT.DSL as DSL
