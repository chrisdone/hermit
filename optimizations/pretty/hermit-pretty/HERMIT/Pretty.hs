module HERMIT.Pretty where

import GhcPlugins hiding (display)

import Control.Monad

import Language.HERMIT.Optimize
import Language.HERMIT.Plugin
import Language.HERMIT.Primitive.Navigation

import Language.Haskell.TH as TH

plugin :: Plugin
plugin = optimize $ \ fns -> after SpecConstr $
    forM_ fns $ \ fn -> at (considerName $ TH.mkName fn) display
