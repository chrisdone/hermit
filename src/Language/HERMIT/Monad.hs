{-# LANGUAGE InstanceSigs #-}

module Language.HERMIT.Monad
          (
            -- * The HERMIT Monad
            HermitM
          , runHM
          , liftCoreM
          , newIdH
          , newTyVarH
    --      , newVarExprH
    --      , newVarH
          , cloneVarH
            -- * Saving Definitions
          , Label
          , DefStash
          , saveDef
          , lookupDef
          , getStash
            -- * Messages
          , HermitMEnv(..)
          , DebugMessage(..)
          , mkHermitMEnv
          , sendDebugMessage
) where

import Prelude hiding (lookup)

import Data.Map

import GhcPlugins hiding (empty)
import MonadUtils       -- from GHC

import Control.Monad
import Control.Arrow

import Language.KURE

import Language.HERMIT.Core
import Language.HERMIT.Context

----------------------------------------------------------------------------

-- | A label for individual defintions.
type Label = String

-- | A store of saved definitions.
type DefStash = Map Label CoreDef

-- | A way of sending messages to top level
newtype HermitMEnv = HermitMEnv { hs_debugChan :: DebugMessage -> HermitM () }

-- | The HERMIT monad is kept abstract.
newtype HermitM a = HermitM (HermitMEnv -> DefStash -> CoreM (KureM (DefStash, a)))

runHermitM :: HermitM a -> HermitMEnv -> DefStash -> CoreM (KureM (DefStash, a))
runHermitM (HermitM f) = f
{-# INLINE runHermitM #-}

-- | Get the stash of saved definitions.
getStash :: HermitM DefStash
getStash = HermitM (\ _ s -> return $ return (s, s))
{-# INLINE getStash #-}

-- | Replace the stash of saved definitions.
putStash :: DefStash -> HermitM ()
putStash s = HermitM (\ _ _ -> return $ return (s, ()))
{-# INLINE putStash #-}

sendDebugMessage :: DebugMessage -> HermitM ()
sendDebugMessage msg = do env <- HermitM $ \ ch s -> return $ return (s, ch)
                          hs_debugChan env msg
{-# INLINE sendDebugMessage #-}

-- | Save a definition for future use.
saveDef :: Label -> CoreDef -> HermitM ()
saveDef l d = getStash >>= (insert l d >>> putStash)
{-# INLINE saveDef #-}

-- | Lookup a previously saved definition.
lookupDef :: Label -> HermitM CoreDef
lookupDef l = getStash >>= (lookup l >>> maybe (fail "Definition not found.") return)
{-# INLINE lookupDef #-}

-- | Eliminator for 'HermitM'.
runHM :: HermitMEnv -> DefStash -> (DefStash -> a -> CoreM b) -> (String -> CoreM b) -> HermitM a -> CoreM b
runHM env s success failure ma = runHermitM ma env s >>= runKureM (\ (a,b) -> success a b) failure
{-# INLINE runHM #-}

----------------------------------------------------------------------------

instance Functor HermitM where
  fmap :: (a -> b) -> HermitM a -> HermitM b
  fmap = liftM
  {-# INLINE fmap #-}

instance Applicative HermitM where
  pure :: a -> HermitM a
  pure = return
  {-# INLINE pure #-}

  (<*>) :: HermitM (a -> b) -> HermitM a -> HermitM b
  (<*>) = ap
  {-# INLINE (<*>) #-}

instance Monad HermitM where
  return :: a -> HermitM a
  return a = HermitM $ \ _ s -> return (return (s,a))
  {-# INLINE return #-}

  (>>=) :: HermitM a -> (a -> HermitM b) -> HermitM b
  (HermitM gcm) >>= f = HermitM $ \ env -> gcm env >=> runKureM (\ (s', a) -> runHermitM (f a) env s') (return . fail)
  {-# INLINE (>>=) #-}

  fail :: String -> HermitM a
  fail msg = HermitM $ \ _ _ -> return (fail msg)
  {-# INLINE fail #-}

instance MonadCatch HermitM where
  catchM :: HermitM a -> (String -> HermitM a) -> HermitM a
  (HermitM gcm) `catchM` f = HermitM $ \ env s -> gcm env s >>= runKureM (return.return) (\ msg -> runHermitM (f msg) env s)
  {-# INLINE catchM #-}

----------------------------------------------------------------------------

-- | 'CoreM' can be lifted to 'HermitM'.
liftCoreM :: CoreM a -> HermitM a
liftCoreM ma = HermitM $ \ _ s -> do a <- ma
                                     return (return (s,a))
{-# INLINE liftCoreM #-}

instance MonadIO HermitM where
  liftIO :: IO a -> HermitM a
  liftIO = liftCoreM . liftIO
  {-# INLINE liftIO #-}

instance MonadUnique HermitM where
  getUniqueSupplyM :: HermitM UniqSupply
  getUniqueSupplyM = liftCoreM getUniqueSupplyM
  {-# INLINE getUniqueSupplyM #-}

instance MonadThings HermitM where
  lookupThing :: Name -> HermitM TyThing
  lookupThing = liftCoreM . lookupThing
  {-# INLINE lookupThing #-}

instance HasDynFlags HermitM where
  getDynFlags :: HermitM DynFlags
  getDynFlags = liftCoreM getDynFlags
  {-# INLINE getDynFlags #-}

----------------------------------------------------------------------------

newName :: String -> HermitM Name
newName name = do uq <- getUniqueM
                  return $ mkSystemVarName uq $ mkFastString name
{-# INLINE newName #-}

-- | Make a unique identifier for a specified type based on a provided name.
newIdH :: String -> Type -> HermitM Id
newIdH name ty = do name' <- newName name
                    return $ mkLocalId name' ty
{-# INLINE newIdH #-}

-- | Make a unique type variable for a specified kind based on a provided name.
newTyVarH :: String -> Kind -> HermitM TyVar
newTyVarH name kind = do name' <- newName name
                         return $ mkTyVar name' kind
{-# INLINE newTyVarH #-}

-- | This gives an new version of a 'Var', with the same info, and a new textual name.
cloneVarH :: (String -> String) -> Var -> HermitM Var
cloneVarH nameMod v =
        let name = nameMod (getOccString v)
            ty   = varType v
        in
          if isTyVar v
           then newTyVarH name ty
           else newIdH name ty
{-# INLINE cloneVarH #-}

----------------------------------------------------------------------------

-- | A message packet.
data DebugMessage  =  DebugTick String
                   |  DebugCore String HermitC Core    -- ^ A postcard.

mkHermitMEnv :: (DebugMessage -> HermitM ()) -> HermitMEnv
mkHermitMEnv debugger = HermitMEnv
                { hs_debugChan = debugger
                }
{-# INLINE mkHermitMEnv #-}

----------------------------------------------------------------------------
