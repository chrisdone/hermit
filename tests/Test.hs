-- note: need to cabal install temporary to run
module Test where

import Control.Monad

import System.Directory
import System.Exit
import System.FilePath as F
import System.IO (Handle, hGetContents, hPutStrLn, hClose)
import System.IO.Temp (withTempFile)
import System.Process hiding (runCommand)

import HERMIT.Driver

type Test = (FilePath, FilePath, FilePath)

-- subdirectory names
golden, dump :: String
golden = "golden"
dump = "dump"

tests :: [Test]
tests = [ ("concatVanishes", "Flatten.hs", "Flatten.hss")
        , ("concatVanishes", "QSort.hs"  , "QSort.hss"  )
        , ("concatVanishes", "Rev.hs"    , "Rev.hss"    )
        , ("evaluation"    , "Eval.hs"   , "Eval.hss"   )
        , ("factorial"     , "Fac.hs"    , "Fac.hss"    )
        -- broken due to Core Parser: , ("fib-stream"    , "Fib.hs"    , "Fib.hss"    )
        , ("fib-tuple"     , "Fib.hs"    , "Fib.hss"    )
        , ("flatten"       , "Flatten.hs", "Flatten.hss")
        , ("hanoi"         , "Hanoi.hs"  , "Hanoi.hss"  )
        , ("last"          , "Last.hs"   , "Last.hss"   )
        , ("map"           , "Map.hs"    , "Map.hss"    )
        , ("mean"          , "Mean.hs"   , "Mean.hss"   )
        -- requires Criterion: , ("nub"           , "Nub.hs"    , "Nub.hss"    )
        , ("qsort"         , "QSort.hs"  , "QSort.hss"  )
        , ("reverse"       , "Reverse.hs", "Reverse.hss")
        ]

fixName :: FilePath -> FilePath
fixName = map (\c -> if c == '.' then '_' else c)

mkTestScript :: Handle -> FilePath -> IO ()
mkTestScript h hss = do
    hPutStrLn h
        $ unlines [ "set-auto-corelint True"
                  , "set-pp-expr-type Show"
                  , "set-fail-hard True"
                  , "load \"" ++ hss ++ "\""
                  , "top ; down"
                  , "display" -- all the bindings
                  , "resume" ]
    hClose h

main :: IO ()
main = do
    pwd <- getCurrentDirectory

    forM_ tests $ \ (dir, hs, hss) -> do
        withTempFile pwd "Test.hss" $ \ fp h -> do
            putStr $ "Running " ++ dir </> hs ++ " - "

            let fixed = fixName (concat [dir, "_", hs, "_", hss])
                gfile = pwd </> golden </> fixed <.> "ref"
                desired = pwd </> dump </> fixed <.> "dump"
                pathp = ".." </> "examples" </> dir

            b <- doesFileExist gfile
            dfile <- if not b
                     then do putStrLn $ "Reference file (" ++ gfile ++ ") does not exist. Creating!"
                             return gfile
                     else return desired

            mkTestScript h hss

            let cmd = unwords $ [ "(", "cd", pathp, ";", "ghc" , hs ] ++ ghcFlags ++
                                [ "-fplugin=HERMIT"
                                , "-fplugin-opt=HERMIT:main:Main:" ++ fp -- made by mkTestScript
                                , "-v0"
                                , "&> " ++ dfile
                                , ")" ]
                diff = unwords [ "diff", "-b", "-U 5", gfile, dfile ]

            -- putStrLn cmd
            (_,_,_,rHermit) <- createProcess $ shell cmd
            exHermit <- waitForProcess rHermit

            case exHermit of
                ExitFailure i -> putStrLn $ "HERMIT failed with code: " ++ show i
                ExitSuccess   -> return ()

            -- putStrLn diff
            (_,mbStdoutH,_,rDiff) <- createProcess $ (shell diff) { std_out = CreatePipe }
            exDiff <- waitForProcess rDiff
            case exDiff of
                ExitFailure i | i > 1 -> putStrLn $ "diff failed with code: " ++ show i
                _ -> maybe (putStrLn "error: no stdout!")
                           (\hDiff -> do output <- hGetContents hDiff
                                         if null output
                                         then putStrLn "passed."
                                         else do putStrLn "failed:"
                                                 putStrLn output)
                           mbStdoutH
