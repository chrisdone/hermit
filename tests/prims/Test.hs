module Main where

main = print "Test"

------------------------ beta reduction ---------------------
beta_reduce_start :: Int
beta_reduce_start = f 1
  where
        f = \ x -> x + 2 :: Int -- is auto-inlined

beta_reduce_end :: Int
beta_reduce_end = 1 + 2

------------------------ case reduction ---------------------
data Foo a = Bar Int Float a | Baz String

case_reduce_start = case bar of
                        Bar x f a -> show x
                        Baz s -> s
    where {-# NOINLINE bar #-}
          bar = Bar 5 2.1 'a'

case_reduce_end = show (5 :: Int)

------------------------ adding and using a rule ---------------------

--{-# NOINLINE capture_me #-}
capture_me :: Int
capture_me = 99

new_rule_start = capture_me

new_rule_end = 99 :: Int

------------------------ fold -----------------------------

double :: Int -> Int
double x = x + x

fold_start :: Int
fold_start = 5 + 5 + 6

fold_end = double 5 + 6
