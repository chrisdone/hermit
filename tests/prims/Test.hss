-- Use an ASCII render (so we can diff)
set-renderer ascii
-- set-pretty-depth 0
-- Tests: beta-reduce, let-subst
{ rhs-of 'beta_reduce_start
  beta-reduce
  let-subst
}
compare-values 'beta_reduce_start 'beta_reduce_end

-- Tests: case-reduce 
{ rhs-of 'case_reduce_start
  let-subst
  case-reduce
}
compare-values 'case_reduce_start 'case_reduce_end

-- Tests: add-rule
add-rule capture 'capture_me
{ consider 'new_rule_start
  any-call (apply-rule capture)
}
compare-values 'new_rule_start 'new_rule_end

-- Tests: fold
{ rhs-of 'fold_start
  any-bu (fold 'double)
}
compare-values 'fold_start 'fold_end
