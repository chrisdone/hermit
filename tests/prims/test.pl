# hacky tests
#
# Assuming that the HERMIT package has already been install
#

sub compileAndRun {
	my ($style,$mod) = @_;
	print "compiling $mod, using ($style)\n";
	$str = "ghc-7.4.1 ${mod}.hs " .
               "    -fforce-recomp -O2 -dcore-lint ";

	if ($style eq "h") {
	    $str .= "-fplugin=Language.HERMIT.Plugin " .
                    "-fplugin-opt=Language.HERMIT.Plugin:main:Main/${mod}.hermit ";
	}

	if ($style eq "i") {
	    $str .= "-fplugin=Language.HERMIT.Plugin " .
                    "-fplugin-opt=Language.HERMIT.Plugin:main:Main/- ";
	}

	system($str);
	print "==> $?\n";
	print "running $mod\n";
	system("./${mod}");
	print "==> $?\n";
}

compileAndRun($ARGV[1],$ARGV[0]);



