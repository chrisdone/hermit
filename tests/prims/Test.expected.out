compiling Test.hs, using (Test.hs,Test.hss resume)
ghc-7.4.1 Test.hs     -fforce-recomp -O2 -dcore-lint -fsimple-list-literals -fplugin=HERMIT -fplugin-opt=HERMIT:main:Main:Test.hss -fplugin-opt=HERMIT:main:Main:resume 
[1 of 1] Compiling Main             ( Test.hs, Test.o )
Loading package ghc-prim ... linking ... done.
Loading package integer-gmp ... linking ... done.
Loading package base ... linking ... done.
Loading package array-0.4.0.0 ... linking ... done.
Loading package deepseq-1.3.0.0 ... linking ... done.
Loading package containers-0.4.2.1 ... linking ... done.
Loading package filepath-1.3.0.0 ... linking ... done.
Loading package old-locale-1.0.0.4 ... linking ... done.
Loading package old-time-1.1.0.0 ... linking ... done.
Loading package bytestring-0.9.2.1 ... linking ... done.
Loading package unix-2.5.1.0 ... linking ... done.
Loading package directory-1.1.0.2 ... linking ... done.
Loading package pretty-1.1.1.0 ... linking ... done.
Loading package process-1.1.0.1 ... linking ... done.
Loading package Cabal-1.14.0 ... linking ... done.
Loading package binary-0.5.1.0 ... linking ... done.
Loading package bin-package-db-0.0.0.0 ... linking ... done.
Loading package hoopl-3.8.7.3 ... linking ... done.
Loading package hpc-0.5.1.1 ... linking ... done.
Loading package template-haskell ... linking ... done.
Loading package ghc-7.4.1 ... linking ... done.
Loading package extensible-exceptions-0.1.1.4 ... linking ... done.
Loading package text-0.11.1.13 ... linking ... done.
Loading package attoparsec-0.10.1.1 ... linking ... done.
Loading package blaze-builder-0.3.1.0 ... linking ... done.
Loading package dlist-0.5 ... linking ... done.
Loading package hashable-1.1.2.3 ... linking ... done.
Loading package transformers-0.2.2.0 ... linking ... done.
Loading package mtl-2.0.1.0 ... linking ... done.
Loading package syb-0.3.6 ... linking ... done.
Loading package time-1.4 ... linking ... done.
Loading package unordered-containers-0.2.1.0 ... linking ... done.
Loading package primitive-0.4.1 ... linking ... done.
Loading package vector-0.9.1 ... linking ... done.
Loading package aeson-0.6.0.2 ... linking ... done.
Loading package ansi-terminal-0.5.5 ... linking ... done.
Loading package data-default-0.4.0 ... linking ... done.
Loading package terminfo-0.3.2.3 ... linking ... done.
Loading package utf8-string-0.3.7 ... linking ... done.
Loading package haskeline-0.6.4.7 ... linking ... done.
Loading package kure-2.1.15 ... linking ... done.
Loading package marked-pretty-0.1 ... linking ... done.
Loading package stm-2.2.0.1 ... linking ... done.
Loading package hermit-0.1 ... linking ... done.
["main:Main:resume","main:Main:Test.hss"]
["main:Main:resume","main:Main:Test.hss"]
("files",["resume","Test.hss"])
0
[including Test.hss]
module main:Main where
  $dShow :: Show [] Char
  main :: IO ()
  beta_reduce_start :: Int
  beta_reduce_end :: Int
  main :: IO ()

module main:Main where
  $dShow :: Show [] Char
  main :: IO ()
  beta_reduce_start :: Int
  beta_reduce_end :: Int
  main :: IO ()

(\ x -> (+) * $fNumInt x (I# 2)) (I# 1)

let x = I# 1 in (+) * $fNumInt x (I# 2)

(+) * $fNumInt (I# 1) (I# 2)

module main:Main where
  $dShow :: Show [] Char
  main :: IO ()
  beta_reduce_start :: Int
  beta_reduce_end :: Int
  main :: IO ()

beta_reduce_start and beta_reduce_end are equal
Linking Test ...
